rate limit: add computed descriptors

kyessenov opened the pull request:
Signed-off-by: Kuat Yessenov <kuat@google.com>

Commit Message: Introduce computed descriptors to rate limits. Refactors the rate limit descriptor interface to accept StreamInfo and RequestHeaderMap. As a general refactor, adds an extension point for alternative rate limit descriptors to register.
Risk Level: low, not enabled unless used
Testing: unit
Docs Changes: yes
Release Notes: yes

kyessenov commented on api/envoy/config/route/v3/route_components.proto:
```diff
@@ -1712,6 +1712,27 @@ message RateLimit {
+
+      oneof expr_specifier {
+        // Expression string, e.g. "connection.requested_server_name".
+        string text = 2 [(validate.rules).string = {min_len: 1}];
```
I decided to skip parsed expr proto for now to avoid introducing a dependency from core APIs on googleapis.

markdroth commented:
I think I need some more context here.  What is this going to be used for, and how do you expect that it will be used?

kyessenov commented:
@markdroth Rate limit descriptor act as categories for requests, since token buckets are applied per category. We have a requirement to use source pod labels as a category. Our extension exposes this information in a filter state, so we need a way to extract this info from the filter state (which means traverse the filter state down to a field).

The current design relies on adding more and more oneof messages here, which we can avoid by just using a generic AST from CEL.

gargnupur commented on api/envoy/config/route/v3/route_components.proto:
```diff
@@ -1712,6 +1712,27 @@ message RateLimit {
+      // The key to use in the descriptor entry.
+      string descriptor_key = 1 [(validate.rules).string = {min_len: 1}];
+
+      oneof expr_specifier {
```
why is this a oneof?

kyessenov replied:
It's anticipating parsed_expr here which would introduce a dependency on expr proto. I can drop it since it's not needed right now.

markdroth replied:
If you think we're going to need the `oneof` later, let's keep it for now.  We can't add it later without breaking the go proto API.

markdroth replied:
Also, I believe that we are actually already using the CEL parsed expression proto in other places in the API, so this is probably not a new external dependency.  You might be able to just add it now.

kyessenov replied:
The dependency is coming from a filter. This is a core proto. Thoughts @htuch?

markdroth replied:
If we don't want to depend on the CEL protos for the xDS core protos, another option might be to make this more extensible by simply adding a `TypedExpressionConfig`.  In the long term, we could maybe even migrate the existing options here to be extensions as well.

@envoyproxy/api-shepherds, thoughts?  I wonder if we should add something to the API review checklist in #14399 that prompts people to consider using `TypedExtensionConfig` instead of a `oneof`?

htuch replied:
From an Envoy binary perspective, are we adding this to core router, or does it only apply to rate limit filter? Hard to tell from config. If core, @kyessenov would be nice to have clarity on when CEL can a reasonable number of the checkmarks in https://github.com/envoyproxy/envoy/blob/master/DEPENDENCY_POLICY.md#new-external-dependencies. Whatever is reasonable, but if this is going to become a core data plane dependency, it's probably good to use this opportunity to check in. CC @envoyproxy/dependency-shepherds 

From an API perspective, I don't feel super strongly, CEL is a reasonably standard thing. One area to have some concern around is that using CEL is similar to regex; we're introducing dynamic evaluation of some expression, which ultimately might have security and inter-request isolation concerns. I don't know if this is in any way actionable, but certainly when communicating in comments/docs its worth pointing out what the runtime implications are.

kyessenov replied:
I believe the runtime change is scoped to global and local rate limit filters. So this is purely a question of a proto dependency, to avoid parsing overhead with a parsed expression-as-proto. Good point about dynamic evaluation, I'll document that. 

Expression evaluation is linear in both input and expression size. The input here is request metadata.

mattklein123 replied:
The fact that rate limit stuff is in the main router is a historical artifact of before we had per-filter route config, etc. None of this should be here and I think there is a tracking issue to move it all out, but I don't think that should block this PR.

kyessenov replied:
OK, I think this forces moving expr package into common from extensions, since it gets pulled in the implementation.

markdroth commented on api/envoy/config/route/v3/route_components.proto:
```diff
@@ -1712,6 +1712,27 @@ message RateLimit {
+    // .. code-block:: cpp
+    //
+    //   ("<descriptor_key>", "<expression_value>")
+    message Expression {
```
Please add a `[#not-implemented-hide:]` comment here.

kyessenov replied:
Implemented.

markdroth commented on api/envoy/config/route/v3/route_components.proto:
```diff
@@ -1742,6 +1763,9 @@ message RateLimit {
       MetaData metadata = 8;
+
+      // Rate limit on a computed expression value.
+      Expression expr = 9;
```
Please add a `[#not-implemented-hide:]` comment here.

kyessenov replied:
Implemented.

kyessenov commented:
The PR is ready for review.

I wonder what I should do about the layering enforcement between core and extensions. It seems that rate limit descriptors are in the core as a historic artifact, so I'd need to pull "Expr" package into `source/common` to match it because it is a dependency from the router. Is that a reasonable approach? Ideally, rate limit descriptors would be another extension point and separated from the router, but I don't see an easy way to achieve that right now.

kyessenov commented:
Bringing this back to attention: I am looking how to resolve the layering violation in API with rate limit descriptors being in the core. There are a couple of ways around it:
1. Add an opaque Any extension to rate limit descriptors, register additional descriptors in rate limit extensions.
2. Keep as-is, move expr-related packages (CEL, other transitive stuff like ANTLR) as a "data-plane" dependency.

Opinions?

mattklein123 commented:
> Opinions?

Definitely not (2). As I already mentioned, the rate limit stuff is only in the core due to legacy reasons. It should all be removed. If you want to take one for the team we can do that (deprecate, move to filter specific route config).

If you aren't willing to do that I think (1) is our best option and we can add extensions there as @rgs1 is doing currently for access log formatters. In thinking about it we probably want an extension point here anyway, so I guess you could just go ahead and add it orthogonally to moving the stuff out of core.

kyessenov commented:
Thanks! Updated the PR to address a new extension point for rate limit descriptors and added computed descriptors as an extension. I did some basic refactoring around rate limit interfaces o make it like other extension points.

markdroth commented:
/lgtm api

kyessenov commented on CODEOWNERS:
```diff
@@ -160,3 +160,5 @@ extensions/filters/http/oauth2 @rgs1 @derekargueta @snowp
 # HTTP Kill Request
 /*/extensions/filters/http/kill_request @qqustc @htuch
+# Rate limit expression descriptor
+/*/extensions/descriptors/expr @kyessenov @lizan
```
@lizan Added as a code owner. Let me know if that doesn't work for you.

lizan replied:
ack, though I think `extensions/descriptors` are too generic, can we have a better name?

kyessenov replied:
We can spell it out as `rate_limit_descriptors`. Would that be better?

kyessenov commented:
Windows failure appears to be identical to Wasm filter build failure (ANTLR).

kyessenov commented:
/retest

kyessenov commented:
/retest

kyessenov commented:
/retest

kyessenov commented:
The coverage checker counts the #ifdef and NOT_REACHED_GCOVR_EXCL_LINE as no coverage for some reason. Any hint to work around it?

kyessenov commented:
/retest

kyessenov commented:
/retest

kyessenov commented:
@lizan Could you please re-approve and merge? There's a merge conflict with https://github.com/envoyproxy/envoy/pull/14588 that I need to fix.

kyessenov commented:
@moderation Hi, do you have a minute to approve deps change in this PR?

moderation commented:
Done @kyessenov. Metadata only deps change
/lgtm deps

mattklein123 reviewed (approved):
Just skimmed, but this is great, thanks.
